### ISSUE TITLE
About Adaptive Quantization

### ISSUE THREAD

**Issue by Carlyx** [CyberAgentAILab/layout-dm#7]
Hi, I just built my own dataset about the real scenes.
And I find that the Kmeans is helpful to make the location more balanced, but it also may lead to the unsuitable relationship.
such as hat and person:
![微信截图_20230515105752](https://github.com/CyberAgentAILab/layout-dm/assets/34021717/9870b9ae-0619-4698-b173-82f84710899d)
![微信截图_20230515105759](https://github.com/CyberAgentAILab/layout-dm/assets/34021717/b10229fb-5a15-44d5-8ed0-f25f663f8f8d)

**Comment by naoto0804** [comment]
It is possible, but I'm surprised to see that the difference is too large. How many bins do you use?

Another reason might be the misunderstanding of bbox format. We use the [center_x, center_y, width, height] format instead of [left, top, width, height].

**Comment by Carlyx** [comment]
I used 365 categories of labels and 9w images (similar to rico25).  And configurations are the same as the given config.yaml in pre-trained files, which means number of bins is 32.
In terms of data processing, I also use the [center_x, center_y, width, height] format.

In addition, I also want to train a layout with a maximum of only 4 elements (M=4). Besides setting max_seq_length as 4,  could you give me some suggestions on what other hyperparameters should be adjusted?

**Comment by naoto0804** [comment]
Thank you for the clarification. Have you checked the values of the center coordinates of each cluster? 

> Could you give me some suggestions on what other hyperparameters should be adjusted?

Hmm, in principle no parameter other than M is affected. But it is possible that some parameters such as num_timesteps (sorry, just a wild guess) can be adjusted because M=4 seems much easier.

**Comment by Carlyx** [comment]
Ok, thank for your patient reply! I have solved the above problems.

And I just have another question.
How can I control the ratio of H to W?
For example, given a category and a tensor 0.5, the model can generate layout with position, H, and W, in which H/W=0.5.

**Comment by naoto0804** [comment]
> How can I control the ratio of H to W?

Although it is not experimentally demonstrated, it is also possible to implement as a weak constraint. Please take a look at A.2 of [our supplementary material](https://openaccess.thecvf.com/content/CVPR2023/supplemental/Inoue_LayoutDM_Discrete_Diffusion_CVPR_2023_supplemental.pdf).

**Comment by naoto0804** [comment]
> Ok, thank for your patient reply! I have solved the above problems.

Out of curiosity, what was the actual reason?

**Comment by Carlyx** [comment]
Well, thanks a lot, I will read the supplementary material.

Actually, speaking of the kmeans, It's about the data processing.  I just used different code with different data processing.
